### PR TITLE
AAD app auth

### LINKED ISSUE #175: Enable AAD Authentication for Apps
AAD Auth needs to be enabled, as per CIS 9.1

Use EasyAuth within app service for this

### PR DESCRIPTION
**Pull request by t-young31**
## Resolves #175

### CODE REVIEW

File: apps/app/locals.tf [hunk 1]
```diff
@@ -19,7 +19,33 @@ locals {
...
   create_repo        = var.app_config.managed_repo != null
 
-  testing_slot_name   = "testing"
+  require_auth = var.app_config.require_auth || var.accesses_real_data
+
+  webapp_name       = "webapp-${replace(var.app_id, "_", "-")}-${var.naming_suffix}"
+  testing_slot_name = "testing"
+  slot_webapp_name  = "${local.webapp_name}-${local.testing_slot_name}"
+
+  webapp_names_requiring_auth = (
```
Comment (jjgriff93) [inline apps/app/locals.tf:28]: What's this doing? Perhaps a comment here as the logic is hard to follow

File: apps/apps.sample.yaml [hunk 2]
```diff
@@ -3,17 +3,14 @@ dash_seedling: # unique ID for the app
   name: Dash Seedling
   description: Hello world app for Dash
   add_testing_slot: false
-  owners: # GitHub usernames of owners to add to repo (for reviews)
-    - octocat: octocat@octocat.com
-  contributors: # contributors to add to repo
-    - octokitten: octokitten@octokitten.com
+  require_auth: false # Should azure active directory authorisation be required?
```
Comment (jjgriff93) [inline apps/apps.sample.yaml:6]:
```suggestion
  require_auth: false # Should Azure Active Directory authentication be required?
```

**Review by jjgriff93 — COMMENTED** [review]
Nice workaround for the issue with v2; still feel like we should look into v1 before we go with this as it could well make this code a lot simpler and mean we don't need to have that hefty json and azapi provider. Hope that's ok and happy to help look into this tomorrow

File: apps/app/platform.tf [hunk 3]
```diff
@@ -71,6 +76,130 @@ resource "azurerm_linux_web_app" "app" {
...
   }
+
+  lifecycle {
+    ignore_changes = [
+      auth_settings_v2
+    ]
+  }
+}
+
+# workaround for https://github.com/hashicorp/terraform-provider-azurerm/issues/20820
```
[the remark this answers is not in the document]
Comment (t-young31) [reply apps/app/platform.tf:87]: yeah... I'm not sure it'd look much nicer because of either the dynamic blocks or copy-paste between webapp/slot resource. happy to go with whatever you think though 🤷🏼

File: apps/app/data.tf [hunk 4]
```diff
@@ -40,7 +40,7 @@ data "azurerm_mssql_server" "feature_store" {
 }
 
 data "azuread_user" "contributors_ids" {
-  for_each            = var.app_config.contributors
+  for_each            = var.app_config.managed_repo != null ? var.app_config.managed_repo.contributors : {}
```
[the remark this answers is not in the document]
Comment (t-young31) [reply apps/app/data.tf:43]: I'm happy to. well, maybe happy is too strong a term 😉

**Comment by jjgriff93** [comment]
Potentially superseded by #212 depending on what we decide

**Comment by t-young31** [comment]
decided on v1 auth (#212 ) based on no azapi call
